pkg/util/grace/grace_wrapper.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grace
+
+import (
+	"k8s.io/klog/v2"
+)
+
+func RunWithGraceSeconds(key, action string, graceSeconds int32, f func() (bool, error)) (bool, error) {
+	return runWithGraceSeconds(DefaultGraceExpectations, key, action, graceSeconds, f)
+}
+
+// Returns:
+// - error: If the passed function itself returns an error, the error is directly returned.
+// - bool: Tells the caller whether a retry is needed.
+// The passed function `f` needs to be idempotent.
+// - The bool value returned by `f` indicates whether the related resources were updated in this call.
+// - If resources were updated, we need to wait for `graceSeconds`.
+func runWithGraceSeconds(e *realGraceExpectations, key, action string, graceSeconds int32, f func() (bool, error)) (bool, error) {
+	modified, err := f()
```

Comment:
shall we execute f() after observing the previous action ? if f() differ from the previous one, reusing the grace expectation may result in incorrect grace time(smaller grace time)